lib/codegen/Decision.cpp

```diff
@@ -18,16 +20,54 @@ void Decision::operator()(DecisionNode *entry, llvm::StringMap<llvm::Value *> su
   }
 }
 
+std::set<std::string> DecisionNode::collectVars() {
+  std::set<std::string> defs, vars;
+  collectDefs(defs);
+  collectUses(vars);
+  for (std::string var : defs) {
+    vars.erase(var);
+  }
+  return vars;
+}
+
+bool DecisionNode::beginNode(Decision *d, std::string name, llvm::StringMap<llvm::Value *> &substitution) {
+  if (isCompleted()) {
+    llvm::BranchInst::Create(cachedCode, d->CurrentBlock);
+    for (std::string var : collectVars()) {
+      phis.lookup(var)->addIncoming(substitution[var], d->CurrentBlock);
+      substitution[var] = phis.lookup(var);
```

Comment:
Is substitution used if we take the early return here?

Reply:
I suppose not, but it doesn't hurt anything. I can remove it and see if the tests still pass if you want though?

Reply:
If the function returns true, the caller should also return. Which would then deallocate the stringmap. so we could delete this if you wanted. I just think it's nicer to keep it for consistency in case the interface changes later.